TASK-45: Authenticated Backend Routes

r-saba opened the pull request:
Closes #169 

Our Routes are secured by either using the newly created `AuthGuard` or the `AuthenticatedUser` class and function in `auth.guard.ts`.

Using  `AuthenticatedUser` we can get the userId directly from the header of the request. This means we no longer should be sending userId as a parameter to any graphQL queries or mutations.

For queries or mutations not taking userId as a parameter, I created the `AuthGuard` class. When placed on an endpoint it checks that the header has a valid access token and continues with the request. If the access token is invalid an error is thrown and the request is not processed.

**Notes**
Queries to our GraphQL server now requires a header with the following field
`Auhtorization: 'Bearer <valid-accesToken>'`

This is what it would look like in apollo studio
![image](https://user-images.githubusercontent.com/25574985/107086388-eef33080-67c7-11eb-8607-5496820b6fb3.png)


Ideally we'd have a script to generate access tokens, I haven't found a pretty solution I'm comfortable pushing at the moment. Feel free to work on this if it interests you.

You can get a valid access token by going to the network tab in your browser and looking at the header of a graphql query made when logged in.

![image](https://user-images.githubusercontent.com/25574985/107086633-485b5f80-67c8-11eb-87f3-89b2f8ddee10.png)

**Running e2e tests**
You will need to add a  `cypress.env.json` file in mull-ui-e2e. Take a look at the docs and resources channel to see what goes inside

r-saba committed:
feat(auth-guard.ts): add authentication parameter decorator

Refs #169

r-saba committed:
feat: add auth guard

Refs #169

r-saba committed:
feat: add authentication to all graphql resolvers

resolvers that use userId as an argument get the userId from a parameter decorator which reads the
ID from the header

Refs #169

r-saba committed:
feat(authguards): use http context instead of gql context

I wasn't using any specific features of the gql context and testing is easier with http context

Refs #169

r-saba committed:
build: add testing utility library as a dev dependency to facilitate mocking http context

Refs #169

r-saba committed:
test: add tests for auth.guard.ts functions

Refs #169

bit172 commented on apps/mull-api/src/app/auth/auth.guard.ts:
```diff
@@ -0,0 +1,33 @@
+import { Observable } from 'rxjs';
+import { environment } from '../../environments/environment';
+
+@Injectable()
+export class AuthGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean | Promise<boolean> | Observable<boolean> {
+    const accessToken = context.switchToHttp().getNext().req.headers.authorization.split(' ')[1];
+    try {
+      return !!jwt.verify(accessToken, environment.jwt.accessSecret);
+    } catch (err) {
+      throw new UnauthorizedException('Unauthorized');
+    }
+  }
+}
+
+export const AuthenticatedUser = createParamDecorator((_, context: ExecutionContext): number => {
+  const accessToken = context.switchToHttp().getNext().req.headers.authorization.split(' ')[1];
+  try {
+    const { id } = jwt.verify(accessToken, environment.jwt.accessSecret) as IAuthToken;
+    return id;
+  } catch (err) {
+    throw new UnauthorizedException('Unauthorized');
```
Nice and clean. I like it.  🚀

bit172 commented on apps/mull-api/src/app/auth/auth.guard.ts:
```diff
@@ -0,0 +1,33 @@
+@Injectable()
+export class AuthGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean | Promise<boolean> | Observable<boolean> {
+    const accessToken = context.switchToHttp().getNext().req.headers.authorization.split(' ')[1];
```
I think you should check if the authorization header is present or bring it in the try block since the response returned is a 500 and not a 401 if it's not there. 
![image](https://user-images.githubusercontent.com/36024638/107105866-9b4c0b80-67f6-11eb-86aa-420cc87434eb.png)

r-saba replied:
Good point! Moved it into the try block so we can control what error gets thrown.

bit172 reviewed (changes requested):
Good work! 🚀 Just a small thing about the guard that I saw, the rest is good. Now that we extract the user Id from the access token, is having the user Id on the frontend obsolete? 🤔

r-saba committed:
fix: move code that gets the access token into a try block to throw the intended error when it fails

this allows us to throw a 401 error instead of a 500 error

Refs #169

r-saba commented:
> Good work! 🚀 Just a small thing about the guard that I saw, the rest is good. Now that we extract the user Id from the access token, is having the user Id on the frontend obsolete? 🤔

Hopefully yes, but I see it's still used in privateRoutes

bit172 reviewed (approved):
LGTM!
- [x] Authenticated routes can only be accessed by authenticated users
- [x] Have a reusable Guard in the backend
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

RGPosadas commented on package-lock.json:
```diff
@@ -10328,22 +10334,14 @@
-          "version": "11.0.0",
-          "resolved": "https://registry.npmjs.org/graphql-upload/-/graphql-upload-11.0.0.tgz",
-          "integrity": "sha512-zsrDtu5gCbQFDWsNa5bMB4nf1LpKX9KDgh+f8oL1288ijV4RxeckhVozAjqjXAfRpxOHD1xOESsh6zq8SjdgjA==",
+          "version": "8.1.0",
```
Why did we downgrade graphql-upload + remove other deps? When I run `npm i`, it brings graphql-upload  back to 11.0.0, and brings back `fs-capacitor` and `isobject`

r-saba replied:
Good question I actually have no idea why, ran `npm i` again and it upgraded to the correct version

r-saba replied:
Done

cristian-aldea replied:
I think i know why, when you do `npm i package` it probably doesn't force resolutions. I ran `npm i` again too and then it finally force resolved my packages.

RGPosadas commented on apps/mull-api/src/app/user/user.resolver.ts:
```diff
@@ -24,17 +27,19 @@ export class UserResolver {
   }
 
   @Mutation(/* istanbul ignore next */ () => User)
+  @UseGuards(AuthGuard)
```
I don't think we need it here for createUser as I just tried to create an account and it told me I was unauthorized.

r-saba replied:
Done

RGPosadas reviewed (changes requested):
Locally, everything works great after taking out the guard on createUser!

However, blocking due to 4 failing mull-ui unit tests (from the pages that you've modified), and the e2e tests are broken as well due to the guards. Example from US-1.1, it fails to create the event with a Fatal Error:
![image](https://user-images.githubusercontent.com/33991121/107122389-85296400-6865-11eb-958e-4d8647649e21.png)

leojrsilao reviewed (changes requested):
Everything looks great with the guards 🚀. 
Noticed as well that there are 4 unit tests failing in the front end and some e2e tests are failing because of the guards.

r-saba committed:
test: update mock queries in tests to match the queries they mock

Refs #169

r-saba committed:
fix: fix package-lock.json changes from previous commit

Refs #169

r-saba committed:
fix: remove authguard from createUser resolver

Refs #169

r-saba committed:
feat: use paramDecorator instead of guard on createEvent mutation

this removes the last resolver in the backend that was receiving userId as a parameter

Refs #169

r-saba committed:
test: update tests for createEvent resolver and service

Refs #169

r-saba committed:
feat(e2e): add dynamic accesstoken generation for e2e tests

Refs #169

r-saba committed:
docs: add documentation for e2e environment variables

Refs #169

r-saba committed:
refactor: change import to use es6 syntax

Refs #169

RGPosadas commented on apps/mull-api/src/app/event/inputs/event.input.ts:
```diff
@@ -3,12 +3,11 @@ import { Field, InputType, Int } from '@nestjs/graphql';
 export class CreateEventInput {
   @Field()
-  title: string;
+  title?: string;
```
Why optional title?

r-saba replied:
Shouldn't be, done

RGPosadas reviewed (commented):
Great work! Tests + linting also pass.
Though please run the backend to also update mull-api's schema.gql as there is still a remnant of hostId.

cristian-aldea commented on package.json:
```diff
@@ -96,6 +96,7 @@
     "@babel/preset-env": "7.9.6",
     "@babel/preset-react": "7.9.4",
     "@babel/preset-typescript": "7.9.0",
+    "@golevelup/nestjs-testing": "^0.1.2",
```
❔ Why was this package added? I see that its used to mock the execution context for nestjs. But I'm wondering if its possible to mock it without a new package.

r-saba replied:
It's a nice testing utility for our project. Especially for mocking context. Here's a snippet from their docs

> This function will create a mock object for you with all sub properties mocked as jest.fn() unless otherwise provided, to allow for easy mocking later on, but more on that later.

I also use this repo as a reference for nestjs testing and that's where I found this utility.  [testing-nestjs](https://github.com/jmcdo29/testing-nestjs)

cristian-aldea commented on apps/mull-api/src/app/auth/auth.guard.spec.ts:
```diff
@@ -0,0 +1,81 @@
+import { createMock } from '@golevelup/nestjs-testing';
```
💡I see that some code still isn't being covered. Might be good to test those lines

![image](https://user-images.githubusercontent.com/23544999/107156347-c98f2f80-694b-11eb-95c3-d0b53b2017b1.png)

r-saba replied:
Assumed since I was testing the function I would get coverage. I found this [github issue](https://github.com/nestjs/nest/issues/1020) and after applying it, it gives us our line coverage while testing the function too.

cristian-aldea reviewed (commented):
Looks good, currently, the application works just fine, and testing and lining are passing. I can't run the e2e tests thou for some reason:
![image](https://user-images.githubusercontent.com/23544999/107156630-83d36680-694d-11eb-84a8-c93ac5e8a167.png)


Just leaving comments for now

cindyslittleplanet reviewed (approved):
Everything passes now, so LGTM! 

- [x] Authenticated routes can only be accessed by authenticated users
- [x] Have a reusable Guard in the backend
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

r-saba committed:
chore(schema.gql): add updated schema.gql file

Refs #169

r-saba committed:
fix: make title of createEventInput mandatory

Refs #169

r-saba committed:
test(authguard): update tests for param decorator based on examples found online

Refs #169

r-saba committed:
fix(linter): disable lint warnings for function used in param decorator tests of auth.guard.spec.ts

Refs #169

r-saba committed:
fix(edit-profile): move back button outside of form

Refs #186

RGPosadas reviewed (approved):
Thanks for addressing all my concerns! LGTM 🚀 

**Acceptance Criteria**

- [x] Authenticated routes can only be accessed by authenticated users
- [x] Have a reusable Guard in the backend
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function

r-saba committed:
fix: change import statement for command.ts to fix typescript errors

Refs #169

leojrsilao reviewed (approved):
LGTM! Addressed my concerns 👍 🚀 

**Acceptance Criteria**

- [x] Authenticated routes can only be accessed by authenticated users
- [x] Have a reusable Guard in the backend
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every component and function
